File: clang/test/APINotes/boundssafety.m
// RUN: rm -rf %t && mkdir -p %t

// RUN: %clang_cc1 -fmodules -fimplicit-module-maps -fmodules-cache-path=%t/ModulesCache -fsyntax-only -fapinotes-modules -I %S/Inputs/Headers -F %S/Inputs/Frameworks -fexperimental-bounds-safety-attributes %s -ast-dump -ast-dump-filter asdf | FileCheck %s

#include "BoundsUnsafeObjC.h"

// CHECK-LABEL: asdf_counted
// CHECK: buf 'int * __counted_by(len)':'int *'
// CHECK-NEXT: len 'int'
// CHECK-NEXT: DependerDeclsAttr {{.*}} 0

// CHECK-LABEL: asdf_sized
// CHECK: buf 'int * __sized_by(size)':'int *'
// CHECK-NEXT: size 'int'
// CHECK-NEXT: DependerDeclsAttr {{.*}} 0

// CHECK-LABEL: asdf_counted_n
// CHECK: buf 'int * __counted_by_or_null(len)':'int *'
// CHECK-NEXT: len 'int'
// CHECK-NEXT: DependerDeclsAttr {{.*}} 0

// CHECK-LABEL: asdf_sized_n
// CHECK: buf 'int * __sized_by_or_null(size)':'int *'
// CHECK-NEXT: size 'int'
// CHECK-NEXT: DependerDeclsAttr {{.*}} 0

// CHECK-LABEL: asdf_ended
// CHECK: buf 'int * __ended_by(end)':'int *'
// CHECK: end 'int * /* __started_by(buf) */ ':'int *'

// CHECK-LABEL: asdf_sized_mul
// CHECK: buf 'int * __sized_by(size * count)':'int *'
// CHECK-NEXT: size 'int'
// CHECK-NEXT: DependerDeclsAttr {{.*}} 0
// CHECK-NEXT: count 'int'
// CHECK-NEXT: DependerDeclsAttr {{.*}} 0

// CHECK-LABEL: asdf_counted_out
// CHECK: buf 'int * __counted_by(*len)*'
// CHECK-NEXT: len 'int *'
// CHECK-NEXT: DependerDeclsAttr {{.*}} IsDeref {{.*}} 1

// CHECK-LABEL: asdf_counted_const
// CHECK: buf 'int * __counted_by(7)':'int *'

// CHECK-LABEL: asdf_counted_nullable
// CHECK: buf 'int * __counted_by(len) _Nullable':'int *'

// CHECK-LABEL: asdf_counted_noescape
// CHECK: buf 'int * __counted_by(len)':'int *'
// CHECK-NEXT: NoEscapeAttr

// CHECK-LABEL: asdf_nterm
// CHECK: buf 'int * __terminated_by(0)':'int *'

File: clang/test/APINotes/bounds-safety.c
// RUN: rm -rf %t && mkdir -p %t
// RUN: %clang_cc1 -fmodules -fimplicit-module-maps -fmodules-cache-path=%t/ModulesCache -fapinotes-modules -fsyntax-only -I %S/Inputs/Headers %s -ast-dump -ast-dump-filter asdf | FileCheck %s

// FIXME: Bounds safety annotations are parsed and stored in the APINotes
// format, but are not yet applied as attributes in the AST upstream since
// __counted_by and related attributes are not yet supported in function
// signatures. Once that support is added, update this test to verify the
// annotations appear on the relevant declarations.

#include "BoundsUnsafe.h"

// CHECK: imported in BoundsUnsafe asdf_counted 'void (int *, int)'
// CHECK: imported in BoundsUnsafe buf 'int *'

// CHECK: imported in BoundsUnsafe asdf_sized 'void (void *, int)'
// CHECK: imported in BoundsUnsafe buf 'void *'

// CHECK: imported in BoundsUnsafe asdf_counted_n 'void (int *, int)'
// CHECK: imported in BoundsUnsafe buf 'int *'

// CHECK: imported in BoundsUnsafe asdf_sized_n 'void (void *, int)'
// CHECK: imported in BoundsUnsafe buf 'void *'

// CHECK: imported in BoundsUnsafe asdf_ended 'void (int *, int *)'
// CHECK: imported in BoundsUnsafe buf 'int *'

// CHECK: imported in BoundsUnsafe asdf_counted_indirect 'void (int **, int)'
// CHECK: imported in BoundsUnsafe buf 'int **'
